[Step2] 동기화메모장 줄라이 릴리

July911 opened the pull request:
찰리 안녕하세요 ! :)
(@kcharliek)
시간 될 때마다 저희 도와주시고, 많은 부분 알려주셔서 정말 감사합니다 ..
덕분에 그래도 얼추 ...구현했습니다.
네임스페이스나, Add 버튼을 눌렀을 때 첫번째 셀로 가는 기능, 키보드 등은 다음 스텝에서 구현해야 할 것 같아요 ..
그래도 CRUD기능이나, 객체간의 이해를 많이 공부해 볼 수 있었던 Step2 였던 것 같습니다.

# 고민했던 점

@escaping 클로저를 통하여, 데이터가 전달되었던 기존 방식을 바꾸어 보았습니다. 

![스크린샷 2022-02-18 오전 12 47 54](https://user-images.githubusercontent.com/90888402/154517952-1a25c8aa-f3e1-4f59-b75f-13768607f97e.jpg)

![스크린샷 2022-02-18 오전 12 47 43](https://user-images.githubusercontent.com/90888402/154517927-3f9b2707-f0e8-46af-8b1d-9a2d2c45e096.jpg)


하지만, 사진 2를 보면 , create 메서드의 종료시점과 updateTableView() 간의 실행 시점이

첫번째 메서드 호출 후 두번째 메서드 호출이 아니라 , 첫번째 메서드 종료 후 두번째 메서드가 실행되야 한다고 생각했습니다. 저의 생각이 맞을까요 ...? 

DispatchQueue 에 넣어주고 세마포어를 통해 통제해주는 방법이나, SplitViewController가 가지고 있는 dataManager 의 dataList가 바뀌었음을 감지해주는 방법 중 어떤 방법이 더 좋은 방법일지 고민됩니다 ! 

# 조언을 얻고 싶은 점

### `reloaData` 를 자주 호출해주는 설계

저희 메모앱은 TextView가 변할 때 마다 불리는 `textViewDidChange` 에서 아래 작업을 실행합니다.

- `CDMemo` 를 바뀐 text로 `setValue` 를 하고
- context에 save하고
- fetch를 통해서 `dataList` 를 최신본으로 유지
- tableView의 `reloadData` 를 호출

따라서 텍스트가 수정될 때 마다 `reloadData` 를 호출하는 형태인데요,, 무거운 작업이라 익히 소문이 나있는 `reloadData` 를 이리 자주 호출해주는 것이 비용측면에서 안 좋을 것이라 생각했지만, 동기화를 위해 불가피하게 이렇게 설계하게 되었습니다. `reloadData` 를 조금이라도 덜 호출하기 위해서, cell에 보이는 영역만 수정될 때만 `reloadData` 를 호출하도록 조건을 걸어주는 방법을 생각해봤는데요..

이에 대한 찰리의 조언을 구합니다🙌🏻

### 아직 구현하지 못한 기능

- [ ]  `nameSpaceForNewCell` 네임스페이스 구현
- [ ]  `createNewMemo` 메서드 호출시 가장 새로운 메모의 Text로 셋팅 + 셀 하이라이트
- [ ]  `TextView` 의 NSAttributeString 설정
- [ ]  메모 화면의 `더보기` 버튼을 선택하면 `액션시트`를 통해 `공유` 및 `삭제` 액션을 표시
- [ ]  `삭제` 를 선택하면 `actionSheet` 표시
- [ ]  `popover` sourceRect 수정
- [ ]  다크모드 구현
- [ ]  `activityView`
    - [ ]  standard service 넉넉히 안뜨는 이슈
    - [ ]  title 지정
    - [ ]  `sourceRect` 수정
    - [ ]  공유기능
- [ ]  키보드
- [ ]  DataManager 프로토콜로 추상화하기

yeahg-dev committed:
feat: Memo Entity 생성 및 saveMemoInContext 구현 #5

July911 committed:
feat: CoreData CRUD 기능 구현 #5

- DataStorage 객체 내부에 CRUD 구현

yeahg-dev committed:
feat: DataManager 타입 구현

- CoreDataProvidable 프로토콜 정의

July911 committed:
refactor: CoreData Entity 이름 변경 #5

- DataManager의 이름을 CoreDataManager로 변경
- Memo Entity의 이름을 CDMemo로 변경

yeahg-dev committed:
feat: 메모 생성 기능 구현 #5

- tappedPlusButton 메서드에 메모 생성 기능 구현
- Entity의 attribute타입 수정을 했는데, CoreData의 영구성으로 인해 충돌이 나서 기존 DataModel 삭제 후 새로운 DataModel 생성 함

yeahg-dev committed:
refactor: 코어데이터모델 변경에 의한 사이드 이펙트 수정 #5

July911 committed:
feat: TableViewCell 삭제 기능 구현 #6

- TrailingContextual TableViewDelegate instance method로 삭제 기능 구현
- DataManager의 Data를 받아와서 View가 그려지게 구현

yeahg-dev committed:
feat: TableViewCell 공유 기능 구현 #6

- SplitViewController에 presentActivityView 구현
- NoteListViewControllerDelegate에 cellToShare 메서드 정의 및 구현
- trailingSwipeActionsConfigurationForRowAt메서드에 shareAction 추가

July911 committed:
feat: Attribute생성 메서드 구현 #5

- Attributes를 body에 넣어주는 delegate구현

yeahg-dev committed:
refactor: fetchAll, save 메서드 리팩터링

yeahg-dev committed:
fix: 메모수정시 0번째 cell없어지는 버그 해결

July911 committed:
refactor: 전체적인 코드 가독성을 위해 리팩토링 #7

- 코딩 컨벤션 수정
- 접근제어자 추가
- 호출순서 순으로 코드 재배치

yeahg-dev committed:
refactor: dataList 접근제어 수정

kcharliek commented:
## 고민했던 점에 대해

맞습니다. create(CoreData Request) & fetch를 마친 이후에 tableView reload를 하는 것이 맞겠지요. 그런데 지금 구현해주신 CoreData 코드는 동기로 동작하기 때문에 문제 없을 것 같기는 합니다만, 이후에 최적화를 고려하면 비동기로 이루어져야합니다. 그렇다면 나중에는, Model의 변화가 이벤트를 만들어 reloadData를 호출하는 그런 흐름을 구상하셔야할 것 입니다.

kcharliek commented:
## 조언을 얻고 싶은 점

알고계신 것 처럼 reloadData는 무거운 요청입니다. 모든 cell을 다시그리는 동작이거든요. 가능하면 실제로 변경된 index만 reloadRow와 같은 인터페이스로 리로드해주는 것이 좋습니다. 

그리고 생각하신 것 처럼 'cell에 보이는 영역만 수정될 때만 reloadData 를 호출하도록' 하는 방법도 좋습니다. 실제로 저도 사용한 경험이 있는 전략이에요. 보이지 않는 cell은 cellForRowAt에서 어차피 갱신될테니까요~

kcharliek commented on CloudNotes/CloudNotes/Controller/NoteDetailViewController.swift:
```diff
@@ -1,6 +1,11 @@
 import UIKit
 
+protocol NoteDetailViewControllerDelegate: AnyObject {
+    func noteDetailViewController(_ viewController: UIViewController, bodyForUpdate body: String)
```
'bodyForUpdate' 라는 네이밍을 보면 이미 NoteDetailView가 언제 update를 해야하지 알고있고 그 타이밍을 제공하고있어요.
NoteDetailView 는 View니까요. 비즈니스로직을 모르는 것이 좋습니다. 그저 '이벤트의 발생' 을 알려주는거죠.
저라면 
``` swift
func noteDetailViewController(_ viewController: UIViewController, didChangeBody body: String)
```

이런 식으로 지었을 것 같아요. 저 delegate에 update를 할지, delete를 할지는 쓰는 녀석이 정하는거죠 .

kcharliek reviewed (approved):
코드 전반적으로 잘 작성해주셨어요 :) 고생 많으셨어요!!!

July911 committed:
feat: NSFetchedController 인스턴스 생성 #5

yeahg-dev committed:
refactor: FetchedResultController가 데이터 제공하도록 수정

July911 committed:
fix: TableView ReloadData 시 데이터가 순서가 안맞는 버그 수정 #7

July911 committed:
refactor: 네이밍 수정 #7

- NoteDetailViewController Delegate 내부 메서드를 didChangeBody로 변경
- 참조순환 해결을 위해 [weak self] 를 추가하여 메모리 해제
